File: hw2/java/Philosopher.java
public class Philosopher extends Thread {

    private String name;
    private double balance;
    private Table table;
    private int seatNo = -1;
    private volatile Order order = null;

    public Philosopher(Table table, String name) {
        this.name = name;
        this.table = table;
        this.balance = 15;
    }

    private void leaveTable() {
        assert this.seatNo != -1;

        this.table.releaseSeat(this.seatNo);
        this.seatNo = -1;
        Logger.log(
            "🚪 Philosopher " +
            this.name +
            " leaves the table with balance $" +
            String.format("%.2f", this.balance)
        );
    }

    private void leaveForInterval(long interval) {
        assert this.seatNo != -1;
        this.leaveTable();

        try {
            Thread.sleep(interval);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public int seatNo(int seatNo, Table table) {
        this.seatNo = seatNo;
        this.table = table;
        return seatNo;
    }

    private Order getOrderFromMenu() {
        assert (this.seatNo != -1);

        if (this.cookedOrder != null) Logger.log(
            "the order for " +
            this.name +
            "is supposed to be null but is " +
            this.cookedOrder.meal()
        );
        assert (this.cookedOrder == null);

        Meal meal = Meal.getRandomMeal();
        Order order = new Order(this.seatNo, meal, this);
        Logger.log("📝 " + this.name + " orders " + meal);
        return order;
    }

    private Pair<Integer, Integer> waitAcquireChopsticks() {
        Logger.log(
            "⏳ " +
            this.name +
            " waits to acquire chopsticks at seat " +
            this.seatNo
        );
        assert (this.seatNo != -1);

        Pair<Integer, Integer> chopsticks = null;
        while (chopsticks == null) {
            chopsticks = this.table.tryGetChopsticks(this.seatNo);
            if (chopsticks == null) {
                Thread.yield();
            }
        }

        Logger.log("🥢 " + this.name + " acquires chopsticks " + chopsticks);
        return chopsticks;
    }

    private void waitForPlacedOrder() {
        while (this.cookedOrder == null) {
            Thread.yield();
            continue;
        }
    }

    private void eat() {
        System.out.println("🍴 " + this.name + " eating " + cookedOrder.meal());
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private void putDownChopsticks() {
        assert this.seatNo != -1;
        this.table.putDownChopsticks(this.seatNo);
    }

    private void payForMeal() {
        this.balance -= this.cookedOrder.price();
        System.out.println(
            "💳 " +
            this.name +
            " paid $" +
            this.cookedOrder.price() +
            " (balance: $" +
            String.format("%.2f", this.balance) +
            ")"
        );
    }

    private boolean hasEnoughBalance() {
        if (this.balance < 0) {
            System.out.println(
                "💸 " +
                this.name +
                " is out of money! Final balance: $" +
                String.format("%.2f", this.balance)
            );
            return false;
        }

        return true;
    }

    private void think(long interval) {
        try {
            Thread.sleep(interval);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private void takeSeatAtTheTable() {
        assert this.seatNo == -1;
        if (this.cookedOrder != null) {
            Logger.log(
                "the order is supposed to be null " +
                this.cookedOrder.meal() +
                " for " +
                this.name
            );
        }
        assert this.cookedOrder == null;
        int seatNo = this.table.requestSeat();
        assert seatNo != -1;
        this.seatNo = seatNo;
        Logger.log(this.name + " takes seat " + seatNo);
    }

    private boolean tryToGiveOrderToRandomWaiter(Order order) {
        assert this.cookedOrder == null;
        assert this.seatNo != -1;

        long startTime = System.currentTimeMillis();
        while (System.currentTimeMillis() - startTime < 2000) {
            Waiter waiter = this.table.getRandomAvailableWaiter();
            if (waiter == null) {
                Thread.yield();
                continue;
            }

            Logger.log(
                "🤝 " +
                this.name +
                " tries to give order to waiter: " +
                waiter.name()
            );

            waiter.takeOrder(order);
            return true;
        }

        return false;
    }

    public void serveOrder(Order order) {
        this.order = order;
    }

    public String name() {
        return this.name;
    }

    @Override
    public void run() {
        while (
            this.hasEnoughBalance() && !Thread.currentThread().isInterrupted()
        ) {
            this.takeSeatAtTheTable();
            this.think((long) (Math.random() * 1000 + 500));

            Order order = this.getOrderFromMenu();
            boolean success = this.tryToGiveOrderToRandomWaiter(order);
            if (!success) {
                Logger.log(
                    this.name +
                    " failed to give order to waiter, leaving table temporarily"
                );
                this.leaveForInterval((long) (1000));
                continue;
            }

            success = this.waitForOrder(order);
            if (!success) {
                Logger.log(
                    "Order failed, was refunded, leaving with refund of $5"
                );
                this.leaveForInterval((long) (1000));
                continue;
            }
            Logger.log(this.name + " got order: " + order.meal());

            this.waitAcquireChopsticks();

            this.eat();
            this.putDownChopsticks();
            this.payForMeal();
            this.leaveForInterval((long) (Math.random() * 1000 + 500));
        }

        System.out.println("👋 " + this.name + " leaving for good.");
    }
}

File: hw2/go/philosopher.go
package main

import (
	"time"
)

type Philosopher struct {
	name              string
	balance           float64
	restaurant        *Restaurant
	seatNo            int
	forkPair          *ForkPair
	orderChan         chan *Order
	orderNotification chan struct{}
}

func newPhilosopher(name string, restaurant *Restaurant) *Philosopher {
	return &Philosopher{
		name:              name,
		balance:           30.0,
		restaurant:        restaurant,
		seatNo:            -1,
		forkPair:          nil,
		orderChan:         make(chan *Order, 1),
		orderNotification: make(chan struct{}, 1),
	}
}

func (p *Philosopher) orderFood() {
	assert(p.seatNo != -1)
	order := &Order{
		orderedBy: p,
		meal:      getRandomMeal(),
		refunded:  false,
	}
	logf("%s orders %s", p.name, order.name())
	p.orderChan <- order
}

func (p *Philosopher) takeSeat() {
	assert(p.seatNo == -1)
	p.seatNo = (int)(p.restaurant.giveSeat())
	logf("%s takes a seat at chair %d", p.name, p.seatNo)
}

func (p *Philosopher) emptySeat() {
	p.restaurant.takeSeatBack((uint)(p.seatNo))
}

func (p *Philosopher) leaveRestaurant() {
	assert(p.seatNo != -1)
	p.emptySeat()
	p.seatNo = -1
	logf("%s leaves the restaurant", p.name)
	time.Sleep(time.Duration(1000))
}

func (p *Philosopher) waitForCookedOrder() *Order {
	logf("%s waits for cooked order", p.name)
	assert(p.seatNo != -1)
	order := <-p.orderChan
	assert(order.orderedBy == p)
	logf("%s receives order %s", p.name, order.name())
	return order
}

func (p *Philosopher) eat(order *Order) {
	assert(p.seatNo != -1)
	time.Sleep(time.Duration(time.Second))
	logf("%s finished eating %s", p.name, order.name())
}

func (p *Philosopher) pay(order *Order) {
	assert(order != nil)
	assert(p.seatNo != -1)
	p.balance -= order.price()
	logf("%s pays for %s, new balance %f", p.name, order.name(), p.balance)
}

func (p *Philosopher) waitForWaiterToTakeOrder() bool {
	logf("%s waits for waiter to take order", p.name)
	assert(p.seatNo != -1)
	timeout := time.After(1 * time.Second)
	select {
	case <-p.orderNotification:
		logf("%s received notification from waiter", p.name)
		return true
	case <-timeout:
		<-p.orderChan
		return false
	}
}

func (p *Philosopher) receiveOrder(order *Order) {
	logf("%s receives order %s at seat %d", p.name, order.name(), p.seatNo)
	assert(order != nil)
	assert(p.seatNo != -1)
	p.orderChan <- order
}

func (p *Philosopher) getRefund() {
	assert(p.seatNo != -1)
	p.balance += 5
}

func (p *Philosopher) notifyOrderSuccess() {
	p.orderNotification <- struct{}{}
}

func (p *Philosopher) waitForChopsticks() {
	logf("%s waiting for chopsticks", p.name)
	assert(p.seatNo != -1)
	p.restaurant.waitGetChopsticks((uint)(p.seatNo))
	logf("%s acquired chopsticks", p.name)
}

func (p *Philosopher) releaseChopsticks() {
	assert(p.seatNo != -1)
	p.restaurant.releaseChopsticks((uint)(p.seatNo))
	logf("%s released chopsticks", p.name)
}

func (p *Philosopher) run() {
	for p.balance > 0 {
		p.takeSeat()
		p.orderFood()
		success := p.waitForWaiterToTakeOrder()
		if !success {
			p.leaveRestaurant()
			continue
		}

		receivedOrder := p.waitForCookedOrder()

		if receivedOrder.isRefunded() {
			p.getRefund()
			p.leaveRestaurant()
			continue
		}

		p.waitForChopsticks()
		p.eat(receivedOrder)
		p.releaseChopsticks()
		p.pay(receivedOrder)
		p.leaveRestaurant()
	}
}
